AVM: EC math

jannotti opened the pull request:
Adds five opcodes to support elliptic curve operations, on the BN254 and BLS 12-381 curves.  These curves are pairing friendly, so opcodes support the normal, expected, operations (`ec_add`, `ec_scalar_mul`, `ec_multi_scalar_mul`) on elliptic curve points, and a `ec_pairing_check` operation on sets of points from associated curves (G1 and G2) allowing for checking zk-proofs.

- [x] Implementations
- [x] Specs
- [x] Tests on standard test vectors
- [x] Benchmarks for costs

Address #3253

fabrice102 commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1560,3 +1560,45 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
+
+| Index | Name | Notes |
+| - | ------ | --------- |
+| 0 | BN254 | BN254 curve, sometimes called BN256 |
```
0 & 3 are not supported for `ec_add`

jannotti replied:
Right, and `ec_pairing_check` only supports 0 and 3.  I suppose I should change things up so that the ops that there is a table of the G1/G2s, and a separate table for `ec_pairing_check`, which will only have 0 and 3 rows from this table.

There's no conceivable operation that would want to take either `BN254` OR `BN254_G1`, right?

jannotti replied:
I've changed this so that only allowable curves are actual G1 and G2 curves of each kind.  No opcode accepts a field like 'BN256` (without the G1 or G2).  The only opcode that _did_ accept such an argument was `ec_pairing_check`. Now, that opcode accepts either the G1 _or_ the G2 form, and expects the named curve's points to be first on the stack, and then the "other" curve's points to be last (top) on stack.

fabrice102 commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1560,3 +1560,45 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
+| 5 | BLS12_381_G2 | G2 of the BLS 12-381 curve |
+
+
+A, B are curve points in G1 group. Each point consists of (X, Y) where X and Y are 256 bit integers, big-endian encoded. The encoded point is 64 bytes from concatenation of 32 byte X and 32 byte Y.
```
```suggestion
A, B are curve points in affine representation: `X` concatenated to `Y`.

| Index | Name | Length of X or Y | Length of A or B |
| ----- | ------ | ---------------- | ----------------- |
| 1 | BN254_G1 | 32 bytes = 256 bits | 64 bytes = 512 bits |
| 2 | BN254_G2 | 64 bytes = 512 bits | 128 bytes = 1024 bits | 
| 4 | BLS12_381_G1 | 48 bytes = 384 bits | 96 bytes = 768 bits |
| 5 | BLS12_381_G2 | 96 bytes = 768 bits | 192 bytes = 1536 bits |


Checks:
* `ec_add` panics if `A` and `B` are not on the curve.
* `ec_add` does not verify that `A` and `B` are in the main prime-order subgroup.
```

Note for some reason https://eips.ethereum.org/EIPS/eip-2537 pads X/Y with 16 bytes for BLS12-381 G1 and with 32 bytes for G2. I don't think we need it.

jannotti replied:
I think this is covered now.  It's annoying to make the table have different headings (they are generated). So the info is in the note field of the table.

fabrice102 commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1560,3 +1560,45 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
+- **Cost**:  BN254=0 BN254_G1=100 BN254_G2=100 BLS12_381=0 BLS12_381_G1=200 BLS12_381_G2=200
+- Availability: v9
+
+A is a curve point in G1 Group and encoded as described in `ec_add`. Scalar K is a big-endian encoded big integer that has no padding zeros.
```
Same checks as for `ec_add`.

Do you want to bound the maximum size of `K`?
Why restricting `K` to have no padding zeros?

jannotti replied:
I don't see a reason for the padding thing. I removed it.

I'll have to benchmark this for large K. I'm guessing the cost is log(K)? If restricted, what is a reasonable restriction, K maybe can be 32 or 64 bytes? (does it just need to be as big as the order of the group?)

jannotti replied:
Commenting before resolving: The limit for scalars is 32 bytes.  This might have to become curve specific someday, if we have curves larger than that.

fabrice102 commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1560,3 +1560,45 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
+- **Cost**:  BN254=1000 BN254_G1=0 BN254_G2=0 BLS12_381=2000 BLS12_381_G1=0 BLS12_381_G2=0
+- Availability: v9
+
+G1s are encoded by the concatenation of encoded G1 points, as described in `ec_add`. G2s are encoded by the concatenation of encoded G2 points. Each G2 is in form (XA0+i*XA1, YA0+i*YA1) and encoded by big-endian field element XA0, XA1, YA0 and YA1 in sequence.
```
```suggestion
G1s are encoded by the concatenation of encoded G1 points, as described in `ec_add`. G2s are encoded by the concatenation of encoded G2 points. Each G2 is in form (XA0+i*XA1, YA0+i*YA1) and encoded by big-endian field element XA0, XA1, YA0 and YA1 in sequence.

Checks:
* `ec_pairing_check` panics if `A` and `B` do not contain the same number of points.
* `ec_pairing_check` panics if any point in `A` or `B` are not on there respective curve.
* `ec_pairing_check` panics if any point in `A` or `B` are not on there respective prime-order subgroup. (Note that this is a different behavior `ec_add` and `ec_scalar_mul`.
```

fabrice102 reviewed (commented):
We also need three more opcodes:
* `ec_map_field_to_curve`: which correspond to BLS12_MAP_FP_TO_G1 / BLS12_MAP_FP_TO_G2 in https://eips.ethereum.org/EIPS/eip-2537
* `ec_is_on_curve`: which checks if a point is on a curve
* `ec_is_in_subgroup`: which checks if a point on the curve is in the subgroup

jannotti commented:
> * `ec_map_field_to_curve`: which correspond to BLS12_MAP_FP_TO_G1 / BLS12_MAP_FP_TO_G2 in https://eips.ethereum.org/EIPS/eip-2537

Added, as `ec_map_to`, so it reads as `ec_map_to BN254_G1`

> * `ec_is_on_curve`: which checks if a point is on a curve

Do we need this if the operations keep checking?  It's not in the EIP. Are there times a contract would want to notice something isn't on curve and do something else.

> * `ec_is_in_subgroup`: which checks if a point on the curve is in the subgroup

Added as `ec_subgroup_check` to match `ec_pairing_check`.  Ilan wasn't able to tell (and neither could I) whether the gnark-crypto subgroup check also implicitly check "on curve" or not. Can an off curve point somehow pass the in subgroup check?

HashMapsData2Value commented:
I'd like to request that ec math for ed25519 also be added. I'd like ed25519 based ring signatures to be verifiable by Algorand smart contracts. https://github.com/HashMapsData2Value/mahber/blob/master/Project_Mahber.pdf

Requested opcodes:
- ed25519 addition (takes two points and adds them)
- ed25519 scalarmult (multiplies point with scalar)
- ed25519 valid point (verify point on curve)
- ed25519 hash to point (hashes input directly to point on curve)

I believe everything needed is already conveniently inside of the internal LibSodium fork.

- [crypto_core_ed25519_add](https://github.com/algorand/go-algorand/blob/master/crypto/libsodium-fork/src/libsodium/crypto_core/ed25519/core_ed25519.c#L26)
- [_crypto_scalarmult_ed25519](https://github.com/algorand/go-algorand/blob/master/crypto/libsodium-fork/src/libsodium/crypto_scalarmult/ed25519/ref10/scalarmult_ed25519_ref10.c#L31)
- [crypto_core_ed25519_is_valid_point](https://github.com/algorand/go-algorand/blob/master/crypto/libsodium-fork/src/libsodium/crypto_core/ed25519/core_ed25519.c#L11)

Regarding the last one, the chain that inspired me to suggest ring signatures (CryptoNote/Monero) [implemented a hash_to_ec here](https://github.com/cryptonotefoundation/cryptonote/blob/8edd998304431c219b432194b7a3847b44b576c3/src/crypto/crypto.cpp#L278)/[here](https://github.com/monero-project/monero/blob/master/src/ringct/rctOps.cpp#L654). But I believe that LibSodium's [crypto_core_ed25519_from_uniform](https://github.com/algorand/go-algorand/blob/master/crypto/libsodium-fork/src/libsodium/crypto_core/ed25519/core_ed25519.c#L66) can be used for this. It [implements the Elligator 2 map](https://elligator.org/hash-to-curve), which accepts a 32 byte value (e.g., the compressed point of a public key like the ed25519_verify opcode) and returns a point on the curve. Someone [wrote a python implementation](https://github.com/bartvm/pyring) (with LibSodium) using LibSodium's crypto_core_ed25519_from_uniform.

barnjamin commented on data/transactions/logic/pairing.go:
```diff
@@ -18,98 +18,743 @@ package logic
+
+// Input: Single byte slice on top of stack containing uncompressed bytes for g1 point
+// Output: Single uint on stack top representing bool for whether the input was in the correct subgroup or not
+func opEcSubgroupCheck(cx *EvalContext) error {
```
Can I please pass an array of points?

barnjamin commented on data/transactions/logic/pairing.go:
```diff
@@ -18,98 +18,743 @@ package logic
+	for i := range scalars {
+		scalars[i].SetBytes(scalarBytes[i*scalarSize : (i+1)*scalarSize])
+	}
+	res, err := new(bn254.G1Affine).MultiExp(g1Points, scalars, ecc.MultiExpConfig{})
```
had to set this locally to have config of `ScalarsMont:true`

jannotti commented:
> I'd like to request that ec math for ed25519 also be added.

I don't think this would be very hard (but different enough that I'd want it in another follow-up PR). However, I wonder if ring signatures can be implemented on these curves instead, if the intent is to do the entire implementation in AVM anyway. (It's possible that would be slower though, I don't know. We'd have to benchmark all these operations.)

HashMapsData2Value commented:
> > I'd like to request that ec math for ed25519 also be added.
> 
> I don't think this would be very hard (but different enough that I'd want it in another follow-up PR). However, I wonder if ring signatures can be implemented on these curves instead, if the intent is to do the entire implementation in AVM anyway. (It's possible that would be slower though, I don't know. We'd have to benchmark all these operations.)

I suspect it would be slower and I think require 2x the storage since points in ed25519 can be represented with only 32 bytes (like the ed25519_verify opcode expects). Benchmarks would be required.

Assuming bn254_G1 the cost for verifying 1 element in a ring signature is ec_scalar_mul(100)+ec_add(10)+ec_scalar_mul(100)+ec_scalar_mul(100)+ec_map_to_g(100)+ec_add(10) + ec_scalar_mul (100) + keccak256(130) = 650 opcost. That's enough to fit in one transaction, which is quite good actually.

![bild](https://user-images.githubusercontent.com/83883690/210617198-2b0fe8a4-883c-48bd-9e16-f9e1e9d4a0fc.png)

jannotti commented:
> EDIT: I saw the opcode costs just now. Assuming bn254_G1 the cost for verifying 1 element in a ring signature is ec_scalar_mul(100)+ec_add(10)+ec_scalar_mul(100)+ec_scalar_mul(100)+ec_map_to_g(100)+ec_add(10) + ec_scalar_mul (100) + keccak256(130) -> 650 cost. That's enough to fit in one transaction, which is quite good actually.

Unfortunately, no, these operations are way more expensive.  I had those numbers in before benchmarking.  After benchmarking, you'll see the only way these things are going to work is in logicsigs which have way more compute allowed.  

The numbers (and comments) may actually be 2x too high right now, and I need to resolve some surprising behavior as I figure out how `scalar_mul` and `pairing_check` scale with multiple inputs.

fabrice102 commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1560,3 +1560,71 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
+| 3 | BLS12_381g2 | G2 of the BLS 12-381 curve. Points encoded as 96 byte X following by 48 byte Y |
+
+
+A and B are curve points in affine representation: X concatenated with Y. Fails if A or B is not in G. Does _not_ check if A and B are in the main prime-order subgroup.
```
```suggestion
A and B are curve points in affine representation: field element X concatenated with field element Y. 

Encoding of a field element `Z` is as follows.
For base field elements (Fp), `Z` is represented as big-endian number and must be lower than the field modulus p.
For the quadratic field extension (Fp2), `Z` is encoded as the concatenation of the individual encoding of the coefficients. For an Fp2 element of the form `Z = Z0 + Z1 i`, where `i` is a formal quadratic non-residue, the encoding of Z is the concatenation of the encoding of `Z0` and `Z1` in this order. (In particular, it fails is `Z0` or `Z1` is larger or equal to p).

The point at infinity is encoded as `(X,Y) = (0,0)`.
Groups G1 and G2 are denoted addively.

Fails if A or B is not in G. 
(A and/or B are allowed to be the point at infinity.)
Does _not_ check if A and B are in the main prime-order subgroup.
```

taken from https://eips.ethereum.org/EIPS/eip-2537, needed to ensure canonicality.

fabrice102 commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1560,3 +1560,71 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
+
+- Opcode: 0xe2 {uint8 curve}
+- Stack: ..., A: []byte, B: []byte &rarr; ..., uint64
+- 1 if all points in A (concatenated points in group G) are paired with the corresponding points of B (concatenated points in the associated group), else 0
```
```suggestion
- 1 if the product of the pairing of each point in A with its respective point in B is equal to the identity element of the target group Gt, else 0
```

jannotti replied:
Thanks, I wrote that before I had any idea what pairing did!

fabrice102 commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1560,3 +1560,71 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
+
+- Opcode: 0xe1 {uint8 curve}
+- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
+- for curve point A and scalar B, return the curve point BA
```
```suggestion
- for curve point A and scalar B, return the curve point BA, the point A multiplied by the scalar B.
```

fabrice102 commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1560,3 +1560,71 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
+- Availability: v9
+
+A is a list of concatenated points, encoded and checked as described in `ec_add`. B is a list of concatenated 32 bytes scalars.
+
```
```suggestion
Contrary to `ec_scalar_mul`, all scalars must be exactly 32-bytes in `ec_multi_exp`.

Note: the name `ec_multi_exp` comes from https://eips.ethereum.org/EIPS/eip-2537 and GNARK, but a more consistent name would be `ec_multi_scalar_mul`.
```

jannotti replied:
I'd rather avoid explicitly referencing the EIP and GNARK directly.  How about:
The name `ec_multi_exp` was chosen to reflect common usage, but a more consistent name would be `ec_multi_scalar_mul`?

Separately. Should we just get rid of `ec_scalar_mul`? You should get the exact same answer using this opcode with one point and one argument. (Internally, surprisingly, we might still call a different gnark routine for the singleton case because it's faster.). This also gets rid of the discussion of the different scalar sizes.

It we do, could we call it `ec_dot_product` and also avoid this discussion?

ahangsu replied:
I up the proposal of renaming to `ec_dot_prod` than `ec_multi_exp`, and probably merge this one against `ec_scalar_mult`, for that is just one entry in vector multiplying another entry in another vec.

We might want to have linear cost model for this one? For the computation varies by the byte length.

jannotti replied:
Currently proposing `ec_multi_scalar_mul` is lower conversation, resolving this one.

fabrice102 commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1560,3 +1560,71 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
+- whether point A is in the main prime-order subgroup of G => {0 or 1}. Program fails if A is not in G at all.
+- **Cost**:  BN254g1=50 BN254g2=11500 BLS12_381g1=5600 BLS12_381g2=7100
+- Availability: v9
+
```
```suggestion

Succeeds for the point at infinity.
```

(I did not check the code yet, but if there is no unit test for that corner case, we definitely need one.)

jannotti replied:
Added, including unit test.
```
func TestSubgroupCheckInfinity(t *testing.T) {
	partitiontest.PartitionTest(t)
	t.Parallel()

	testAccepts(t, "int 64; bzero; ec_subgroup_check BN254g1;", pairingVersion)
	testAccepts(t, "int 128; bzero; ec_subgroup_check BN254g2;", pairingVersion)
	testAccepts(t, "int 96; bzero; ec_subgroup_check BLS12_381g1;", pairingVersion)
	testAccepts(t, "int 192; bzero; ec_subgroup_check BLS12_381g2;", pairingVersion)
}

```

fabrice102 commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1560,3 +1560,71 @@ For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `bo
+- **Cost**:  BN254g1=1700 BN254g2=11000 BLS12_381g1=5600 BLS12_381g2=43000
+- Availability: v9
+
+BN254 points are mapped by the SVDW map. BLS12-381 points are mapped by the SSWU map. G1 element inputs are encoded single big-endian byte-array of length <= n.  G2 inputs are encoded as concatenated n-byte big-endian encoded integers. n == 32 for BN254 and 48 for BLS12-381. No input may exceed the modulus of the group. 
```
```suggestion
BN254 points are mapped by the SVDW map. BLS12-381 points are mapped by the SSWU map. 

G1 element inputs are base field elements and G2 element inputs are quadratic field elements, with the same encoding rules as defined in `ec_add`.
In particular, no input coefficient can exceed the modulus of the group.
```

jannotti replied:
`ec_add` takes G1 points, which are two concatenated 32 big-endian integers.  But here, we take the "base field elements" which for G1, are just single integers. So it does not seem helpful discuss the "same encoding rules as `ec_add`".

In fact, there is quite a departure from those rules, even when you consider the elements themselves. Here, at least for G1 groups, we allow short bytes sequences like 0x0108, and interpret is as 264 (256+8). We don't mind that the input is not 32 bytes long. Since `ec_add` effectively takes two such elements, it _demands_ the input is 64 bytes - two concatenated 32 byte values.

For `ec_map_to` applied to G2 fields, we can not allow such conveniences, since the base field is itself quadratic.  They must be 64 or 94 bytes (2 32s, or 2 48s).  That why I talk about "<= n" for G1 inputs, and exactly n for G2 inputs.

fabrice102 reviewed (commented):
Made some comments on the documentation only.
I'll review the code later.

fabrice102 commented on data/transactions/logic/pairing_test.go:
```diff
@@ -16,14 +16,543 @@
+	g1point := bn254RandomG1()
+	g1bytes1 := fmt.Sprintf("byte 0x%s\n", hex.EncodeToString(bn254G1ToBytes(&g1point)))
+	g1point = bn254RandomG1()
+	g1bytes2 := fmt.Sprintf("byte 0x%s\n", hex.EncodeToString(bn254G1ToBytes(&g1point)))
```
It might be somewhere else, but I think it's important to test also all the corner cases:
* wrong length for first point
* wrong length for second point
* right length but not on curve for first point
* same for second point
* addition with first point = point to infinity (done already)
* addition with second point = point to infinity
* addition with both point of infinity
* addition of points that are on the curve but not in the subgroup.

fabrice102 commented on data/transactions/logic/pairing_test.go:
```diff
@@ -16,14 +16,543 @@
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	// Parameterize this to check all curves
```
I would add similar corner cases as above for ec_add, as well as add scalar = r (the order) that should panic

fabrice102 commented on data/transactions/logic/pairing_test.go:
```diff
@@ -16,14 +16,543 @@
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	t.Run("bn254", func(t *testing.T) {
```
I would add all the corner cases as above, including different lengths of arrays.

fabrice102 commented on data/transactions/logic/pairing_test.go:
```diff
@@ -16,14 +16,543 @@
+		g1bytes := fmt.Sprintf("byte 0x%s\n", hex.EncodeToString(bn254G1sToBytes(g1points)))
+		g2bytes := fmt.Sprintf("byte 0x%s\n", hex.EncodeToString(bn254G2sToBytes(g2points)))
+
+		testAccepts(t, g1bytes+g2bytes+`ec_pairing_check BN254g1`, pairingVersion)
```
I did not realize that you can specify a group G1/G2 for pairing and that switches the two arguments.
I think this is a bit unusual (though not a security issue).
Would is be possible to only allow `BN254` not `BN254g1/g2` and then requires g1 is always first?

jannotti replied:
That's what I had, but creates a bunch of documentation and wording questions because "BN254" is not a curve.  We were all annoyed at need to say that some of the opcodes worked on the "...G1" curves/groups, but not the unsuffixed names, etc

So do we create to different sets of constants? The g1/g2 table, which might be called "curves" and then the table that does not have the suffixes and call the "curve families"?  

Seems like this sidesteps all of that and only talks about the actual curves, the G1/G2s.

shaih replied:
I cannot imagine what questions can possibly arise by insisting on g1-then-g2 ordering.

In general I like the proposal of calling it "BN12-384 family" when referring to the entire system, and talk about G1, G2 when needing to specify a particular group. If there is ever a need to talk about the curves themselves (which I hope there isn't), then call them "curve 1" and "curve 2"

ahangsu commented on data/transactions/logic/pairing.go:
```diff
@@ -18,98 +18,766 @@ package logic
-	res := new(bn254.G1Affine).Add(&a, &b)
-	resBytes := bN254G1ToBytes(res)
+	kBytes := cx.stack[last].Bytes
+	if len(kBytes) > 32 {
```
nit: consider using const `scalarSize`

ahangsu commented on data/transactions/logic/pairing.go:
```diff
@@ -18,98 +18,766 @@ package logic
-	cx.stack[prev].Bytes = resBytes
-	return nil
+	cx.stack[prev] = boolToSV(ok)
+	return err
```
one question: if there is an erroring in pairing check, do we want to hide the erroring by casting existing err to an `ok = False`? I thought we might want to undistinguish following 2 cases:
- not pairing succeed
- pairing inputs are not valid

or in a sense, a failing pairing check should not fail the teal program?

jannotti replied:
I think the intent is to fail the program since an error in the pairing check call is only from misformatted input, so it's probably programmer error. I guess most of the time, returning 0 would be ok (because most code will be written to "do nothing" when pairing fails), but it seems better to just reject the whole thing.

ahangsu replied:
yea that sounds fair to me, I just dug into the implementation, and realizing that the only erroring cases in our implementation and gnark implementation are related with wrong byte size (byte array len 0 or not aligned with 32 bytes), so I agree that it is falling in your category of wrong format input, which should fail anyways tho.

fabrice102 commented on data/transactions/logic/pairing.go:
```diff
@@ -18,98 +18,766 @@ package logic
+// Output: Single uint at top representing bool for whether pairing of inputs was identity
+func opEcPairingCheck(cx *EvalContext) error {
+	/*
+	   Q: Should pairing fail is the supplied points are not canonical?
```
Yes, it should fail.

fabrice102 commented on data/transactions/logic/pairing.go:
```diff
@@ -18,98 +18,766 @@ package logic
+func opEcPairingCheck(cx *EvalContext) error {
+	/*
+	   Q: Should pairing fail is the supplied points are not canonical?
+	   Q: Should pairing fail if there are no inputs?
```
Unclear. https://eips.ethereum.org/EIPS/eip-2537 rejects empty inputs.
Using empty inputs is most likely a mistake from the programmers I would guess... Though mathematically, it makes sense to accept empty inputs.

jannotti replied:
Resolved in favor of EIP compatibility (failing on empty).

ismael-elatifi commented:
Hello @jannotti ,
I want to implement Algorand smart contracts which need to verify ZK-SNARK proofs.
I think this work should enable that.
Do you know approximately when ZK-SNARK verification on-chain should be possible on test net ?

jannotti commented:
> I think this work should enable that.

I think so too. That is the intention. If you'd like to try it out and give feedback to ensure that's true, we'd love that.  You can use `sandbox` to run a local network using this branch.

> Do you know approximately when ZK-SNARK verification on-chain should be possible on test net ?

Once the PR is accepted, it will go to test net wth the next consensus update. They seem to happen every few months, and the last one was about a month ago. So, optimistically, it might be on test net in a couple months. But that assumes we get the PR in soon. Having direct feedback on whether the opcodes here are sufficient for a real use case would be great! What proving system do you intend to use?

jannotti commented on data/transactions/logic/README.md:
```diff
@@ -373,6 +373,12 @@ return stack matches the name of the input value.
+| `ec_add g` | for curve points A and B, return the curve point A + B |
+| `ec_scalar_mul g` | for curve point A and scalar B, return the curve point BA, the point A multiplied by the scalar B. |
+| `ec_pairing_check g` | 1 if the product of the pairing of each point in A with its respective point in B is equal to the identity element of the target group Gt, else 0 |
+| `ec_multi_exp g` | for curve points A and scalars B, return curve point B0A0 + B1A1 + B2A2 + ... + BnAn |
```
I agree. Do you like `ec_dot_product` or `e?_inner_product`. I want to maintain the additive naming of `ec_add` and `ec_mul`?

Can you find any other places that show a clear preference for `dot` or `inner` product? (or another additive name?)

jannotti replied:
Some quick googling seems to indicate that people only like to use "dot product" when the vectors are real numbers, and "inner product" is used when the vectors are something more interesting, like ec points.

Does that fit your understanding and uses you've seen?

ahangsu replied:
yea, was also doing the naming search... I can only say someone explained me with term inner product, but I haven't really seen it closely in documentations. Folks refer to same thing with different names... up until we see the formula.

Both make sense to me, while inner feels better to me, for my search of inner leads to eprint.iacr.org/2022/596.pdf on chapter 4, which is the linear combination we are looking for.

ahangsu replied:
oh actually, pause a sec. maybe multi-scalar-mul? see gnark note https://hackmd.io/@gnark/eccbench here and https://hackmd.io/@tazAymRSQCGXTUKkbh1BAg/Sk27liTW9, I would slightly prefer MSM now, for it is summing multiple scalar-multiplication...

my bad for proposing yet another naming...

jannotti replied:
Ok, how about `ec_multi_scalar_mul`, which compares nicely to `ec_scalar_mul`?

ahangsu replied:
sounds nice, I upvote this choice.

jannotti commented on go.mod:
```diff
@@ -36,14 +36,14 @@ require (
 	github.com/olivere/elastic v6.2.14+incompatible
 	github.com/sirupsen/logrus v1.8.1
-	github.com/spf13/cobra v1.3.0
+	github.com/spf13/cobra v1.5.0
```
I have to go through the `go.mod` and see if I can trim it down. I'll leave this convo open until I do.

jannotti replied:
I was able to remove some of the changes, but this one comes back when I `make tidy`. I have to say I do not understand `go.mod`. I don't know where the version for a package can be forced in this way.

ahangsu replied:
I was trying locally, and here's what I learnt from golang package management...

I tried `go get -u github.com/spf13/cobra@v1.3.0` to roll back to cobra 1.3.0, and ran `go mod tidy`, the dependency changes in:
1. github.com/cpuguy83/go-md2man/v2
2. github.com/inconshreveable/mousetrap

are still there, and I have to manually use `go get` and `go mod tidy` to pin the versions...

I don't think I (will) understand `go.mod` well either...

But probably it is not suggested to directly change the `go.mod` file, but rather use `go get` and `go mod tidy`?

jasonpaulos replied:
I think this change is forced because gnark-crypto v0.8.0 uses this version of cobra.

jannotti replied:
That would explain it. I'm going to consider this resolved then.

jannotti commented on data/transactions/logic/TEAL_opcodes.md:
```diff
@@ -1644,3 +1644,90 @@ Fields
+- Bytecode: 0xe3 {uint8}
+- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
+- for curve points A and scalars B, return curve point B0A0 + B1A1 + B2A2 + ... + BnAn
+- **Cost**:  BN254g1=80 + 3 per 32 bytes of B; BN254g2=180 + 9 per 32 bytes of B; BLS12_381g1=140 + 4 per 32 bytes of B; BLS12_381g2=350 + 18 per 32 bytes of B
```
Correct, this is vastly underpriced (I think 1000x less?) so I could do tests, but now that we have logicsig pooling, I should be able to revamp the tests using proper costs.

ahangsu replied:
got it, was thinking about introduction of logicsig pooling, so I should probably revisit the cost later.

Resolving this thread now.

ahangsu commented on ledger/apptxn_test.go:
```diff
@@ -2156,6 +2156,8 @@ func TestInnerCreatedAppsAreCallable(t *testing.T) {
 	t.Parallel()
 
 	genBalances, addrs, _ := ledgertesting.NewTestGenesis()
+	l := newTestLedger(t, genBalances)
+	defer l.Close()
```
I am not too familiar with testing tools in ledger, but this seems like a reasonable addition by context... not important to this PR I guess

jannotti replied:
I'm going to remove it.  It looks entirely extraneous. Thanks.

ahangsu reviewed (commented):
I only have `eval_test` and `pairing_test` pending on my read stack.  The implementation seems reasonable by using gnark API.

jasonpaulos commented on data/transactions/logic/pairing.go:
```diff
@@ -17,99 +17,876 @@
+	if big.Cmp(bn254Modulus) >= 0 {
+		return bn254fp.Element{}, fmt.Errorf("field element %s larger than modulus %s", &big, bn254Modulus)
+	}
+	return *new(bn254fp.Element).SetBigInt(&big), nil
```
nit: if not compiled away, this also seems like an unnecessary use of `new`

jannotti replied:
I tried this one too, since less was going on here, I had hopes that a stack variable would not escape. But it didn't so allocs unchanged.

jasonpaulos reviewed (commented):
Still poking through tests, but here are my comments so far

jannotti commented on data/transactions/logic/pairing.go:
```diff
@@ -17,99 +17,876 @@
+	if len(kBytes) > scalarSize {
+		return fmt.Errorf("ec_scalar_mul scalar len is %d, exceeds 32", len(kBytes))
+	}
+	k := new(big.Int).SetBytes(kBytes)
```
I think I once convinced myself that the second form is just as bad, because it's really calling a function like: `SetBytes(&big, b)`, and the Go compiler therefore assumes that `&big` escapes, and must be heap allocated, even though you don't see it. So it's not that the `new` gets optimized away, it's that the declaration on the stack gets "pessimized" up to an allocation.

I will write a benchmark to convince myself.

jannotti replied:
Yes, it gets pessimized. I ran my benchmark for `ec_scalar_mul` with `b.ReportAllocs()` and found that there are 13 allocations per loop regardless of how it's written.

jasonpaulos commented on data/transactions/logic/pairing_test.go:
```diff
@@ -16,14 +16,876 @@
+	// - any of the "EC" constants. The first three fields have different
+	// costs, that depend on the length of the input
+
+	fmt.Println("ECGroups", EcGroups)
```
nit: is the println meant to be here?

jannotti replied:
nope!

jasonpaulos commented on data/transactions/logic/pairing_test.go:
```diff
@@ -16,14 +16,876 @@
+				}
+				rand.Read(kbytes)
+				for r := 0; r < b.N; r++ {
+					bls12381G2MultiExpSmall(g2points, kbytes)
```
Is it worth calling `b.ResetTimer()` before the loop?

It's also not clear to me why some code includes the `g1points/g2points` setting and `rand.Read(kbytes)` call inside of the loop over `b.N` and some does not.

jannotti replied:
I don't think there's any point in resetting - the only goal of this benchmark is to find out when one function gets slower than the other. Since they both have the randomness at the beginning, they should have the same overhead, and the cutoff will still be accurate.

Except, and I think this goes to your second question, it looks like I forgot to rerandomize the points for G1 large.  I'll add that.  I will also tuck the rerandomization (for all calls) inside the loop, since I've seen some of these function perform better on repeated runs against the same values.

jannotti commented on data/transactions/logic/fields.go:
```diff
@@ -676,6 +676,74 @@ var EcdsaCurves = FieldGroup{
+	{BLS12_381g2, "G2 of the BLS 12-381 curve. Points encoded as 96 byte X following by 48 byte Y"},
+}
+
+func ecGroupSpecByField(c EcGroup) (ecGroupSpec, bool) {
```
This is probably an outdated comment, the group spec is used now, to pick the group being operated on.

ismael-elatifi commented:
> If you'd like to try it out and give feedback to ensure that's true, we'd love that. You can use `sandbox` to run a local network using this branch.

Unfortunately I don't have enough cryptographic background to implement ZK-SNARK verifier code with TEAL directly, but I hope there will be PyTEAL wrappers added to verify common ZK-SNARK proofs (like the [Tornado Cash Core Deposit Circuit](https://tornadocash.vn/docs/general/circuits/core-deposit-circuit/)).  

Actually I asked a question on the Algorand Forum and I got [this reponse](https://forum.algorand.org/t/is-avm-teal-compatible-with-verification-of-zk-snark-proof/10071/3) from @HashMapsData2Value that pointed to this PR.

> What proving system do you intend to use?

Based on @HashMapsData2Value answer and what is mentionned in this [Tornado Cash page](https://tornadocash.vn/docs/general/circuits/zk-circuits/#snarks-and-groth16), it is GROTH16.   

But actually I don't care about the specific implementation, all I care about is to have the necessary building blocks to make the Tornado Cash anonymous withdrawal use case possible on Algorand (with on-chain ZK-SNARK proof verification).

jannotti commented on data/transactions/logic/doc.go:
```diff
@@ -337,55 +348,70 @@ var opDocExtras = map[string]string{
+		"Does _not_ check if A and B are in the main prime-order subgroup.",
+
+	"ec_scalar_mul":    "A is a curve point encoded and checked as described in `ec_add`. Scalar B is interpreted as a big-endian unsigned integer. Fails if B exceeds 32 bytes.",
+	"ec_pairing_check": "A and B are concatenated points, encoded and checked as described in `ec_add`. A contains points of the group G, B contains points of the associated group (G2 if G is G1, and vice versa). Fails if A and B have a different number of points, or if any point is not in its described group or outside the main prime-order subgroup - a stronger condition than other opcodes AVM valu ...
```
Can anyone else weigh in?  I'd like to remove the discussion of length limits.  They apply universally.

jannotti replied:
I'll leave them in for now, but they could be removed in a little cleanup PR some time.

bbroder-algo reviewed (approved):
Audited code to provide ZK verification capabilities in AVM.

jasonpaulos reviewed (approved):
Looks good to me

ismael-elatifi commented:
Hi @jannotti,
Well done for the merge of this important work !  
Will there eventually be tests proving that we can actually implement ZK proof verification with AVM/TEAL ?  
They would also serve as a great example for people wanting to do the same 🙂
